mitosheet/mitosheet/step_performers/graph/graph.py

```diff
@@ -0,0 +1,203 @@
+            else []
+        )
+        x_axis_column_headers = prev_state.column_ids.get_column_headers_by_ids(
+            sheet_index, x_axis_column_ids
+        )
+
+        # Get the y axis params, if they were provided
+        y_axis_column_ids = (
+            graph_creation["y_axis_column_ids"]
+            if graph_creation["y_axis_column_ids"] is not None
+            else []
+        )
+        y_axis_column_headers = prev_state.column_ids.get_column_headers_by_ids(
+            sheet_index, y_axis_column_ids
+        )
+
+        # Find the height and the width, defaulting to fill whatever container its in
+        graph_rendering_keys = graph_rendering.keys()
+
+        height = graph_rendering["height"] if "height" in graph_rendering_keys else "100%"
+        width = graph_rendering["width"] if "width" in graph_rendering_keys else "100%"
+
+        # Create a copy of the dataframe, just for safety.
+        df: pd.DataFrame = prev_state.dfs[sheet_index].copy()
+        df_name: str = prev_state.df_names[sheet_index]
+
+        if len(x_axis_column_ids) == 0 and len(y_axis_column_ids) == 0:
+            fig = ''
+            html_and_script = {'html': '', 'script': ''}
+            graph_generation_code = ''
+        else: 
+            fig = get_plotly_express_graph(
+                graph_type,
+                df,
+                safety_filter_turned_on_by_user,
+                x_axis_column_headers,
+                y_axis_column_headers,
+            )
+
+            # Get rid of some of the default white space
+            fig.update_layout(
+                margin=dict(
+                    l=0,
+                    r=0,
+                    t=30,
+                    b=30,
+                )
+            )
+
+            html_and_script = get_html_and_script_from_figure(fig, height, width)
+
+            graph_generation_code = get_plotly_express_graph_code(
+                graph_type,
+                df,
+                safety_filter_turned_on_by_user,
+                x_axis_column_headers,
+                y_axis_column_headers,
+                df_name,
+            )
+
+        post_state.graph_data_json[str(sheet_index)] = {
+            "graphParams": {
+                "graphPreprocessing": graph_preprocessing,
+                "graphCreation": graph_creation,
+                "graphStyling": {},
+                "graphRendering": graph_rendering,
+            },
+            "graphGeneratedCode": graph_generation_code,
+            "graphHTML": html_and_script["html"],
+            "graphScript": html_and_script["script"],
+        }
+
+        return post_state, None
+
+    @classmethod
+    def transpile(  # type: ignore
+        cls,
+        prev_state: State,
+        post_state: State,
+        execution_data: Optional[Dict[str, Any]],
+        graph_preprocessing: Any,
+        graph_creation: Any,
+        graph_styling: Any,
+        graph_rendering: Any,
```

Comment:
In all functions here, can we type these as Dict[str, Any]?

Reply:
Also our standard convention is to put execution_data last, right? I might be wrong about that in which case ignore.